_posts/2018-03-15-integrating-redux-and-socketio.markdown

```diff
@@ -0,0 +1,87 @@
+
+To keep that from occurring, their workflow would involve editing a draft in another editor such as google docs and once their article is ready, copy it over to our writer app. The goal of this project was to implement locking mechanism to only allow one editor per article. I had to come up with an elegant technical implementation for this problem. Here's how I went about it.
+
+<!-- more -->
+
+In order to fix that problem, we decided to build an article locking mechanism so only one of the editors on the team can start editing at once. One of the requirements for this new feature was for our editors to see when someone else started editing without refreshing the page. In order to fulfill this, we needed to implement a system to push events from the server to clients.
+
+Based on the requirements presented, I looked at potential solutions for this. Right away, the HTML5 WebSocket API seemed to be the perfect solution to keep all clients synced in realtime, however a few issues arose. For one, many proxies and firewalls block WebSocket connections so it's not always an available option for clients.
+
+At that point, we had already started the process of converting the app from using Backbone + jade to a modern javascript stack. Using React for UI components and Redux for managing app state. So whatever solution I came up with, it had to play well with redux. Redux emerged as the industry standard way implementing one-way data flow into your app, taking actions (events).  If you’re familiar with the redux architecture you know it comprises a single source of truth for state known as the store. Actions, action creators, reducers. (You can read this excellent [intro to react]() blog post to familiarize yourself with the topic if you aren't).
+
+![/images/2018-03-15-integrating-redux-and-socketio/example.gif](/images/2018-03-15-integrating-redux-and-socketio/example.gif)
+
+For server/client, I ended up using socket.io, a battle tested library for two-way real time communication channel. At that point, we were facing another challenge, how do we integrate socket.io in the react/redux architecture we just designed. also written in javascript for backend and clients combine the two. All that’s left to do is to integrate the two libraries.
+
+Here's a simplified version of the helper function that I wrote to broadcast redux actions over the socket to other connected clients via our backend service.
+
+```javascript
+// From client
+import io from 'socket.io-client'
+import { messageTypes } from './messageTypes'
+
+let socket = io(rootURL)
+
+// Helper to emit a redux action to our websocket server
+const emitAction = (actionCreator) => {
+  return (...args) => {
+    // This return the action object which gets sent to our backend
+    // server via the socket connection
+    const result = actionCreator.apply(this, args)
+    socket.emit(result.key, {
+      ...result.payload,
+      type: result.type
+    })
+    return result
+  }
+}
+```
```

Comment:
Curious whether the event ingestion part is in scope for this post, because that would probably be the most interesting part to me as a reader new to real-time stuff.

But as a reader who has used WebSockets, the question for me would be what happens in the unhappy path, like if there's contention and my client doesn't obtain the lock on the article?

Lastly, it might be helpful to call out the pattern of making this a [decorator function](https://leanpub.com/javascriptallongesix/read#decorators) for folks who want to explore this pattern further.